scipy/stats/_bws_test.py

```diff
@@ -0,0 +1,230 @@
+import numpy as np
+from collections import namedtuple
+from ._axis_nan_policy import _axis_nan_policy_factory
+
+
+def calculate_bws_sum(ranks, nm):
+    '''Calculate the sum for the Baumgartner-Weiss-Schindler statistic'''
+    sum = 0
+    size = len(ranks)
+    factor = 1 / (size + 1)
+    for i, r in enumerate(ranks):
+        # index & ranks are 0-based, but we want 1-based
+        i1, r1 = i + 1, r + 1
+        sum += (r1 - nm / size * i1) ** 2 / (i1 * factor * (1 - i1 * factor))
+    return sum
+
+
+def bws_statistic_from_ranks(ranks, n, m):
+    '''Calculate the Baumgartner-Weiss-Schindler statistic from a set of ranks of n values from one sample.'''
+    nm = n + m
+
+    # Calculate the BWS statistic from a set of ranks of n values from one sample
+    # and m values from another sample.
+    b_x = calculate_bws_sum(ranks[:n], nm) / (m * nm)
+    b_y = calculate_bws_sum(ranks[n:], nm) / (n * nm)
+    b = (b_x + b_y) * 0.5
+
+    return b
+
+
+def shuffle(a, b):
```

Comment:
There is so much recursion and array resizing here that I can't follow what is happening. It seems like you are recreating the Python built-ins `zip` and `itertools.product`. Why generators? Why concatenate?

Reply:
Hi, @ilayn I adapted from here: https://math.stackexchange.com/questions/4679930/python-realization-of-permutation-of-2-or-more-groups-while-keeping-the-ordering/4679936#4679936

Reply:
I am afraid we cannot do this. This code's license is not compatible with SciPy's license (CC-BY-SA is incompatible with BSD-3-clause). So this part needs to be overhauled.

Reply:
> I am afraid we cannot do this. This code's license is not compatible with SciPy's license (CC-BY-SA is incompatible with BSD-3-clause). So this part needs to be overhauled.

Wait, but this is quite to the original implementation, this is also the question that I initialized. What should we do atm?

Reply:
Just seeing you asked the question yesterday. Then the easy thing that you can try is just ask permission to use in SciPy under our BSD-3-clause. Then in this code add a comment saying the code if from there and we got permission. (There are examples in our codebase you can look at where we did that if you want an example on how to write that.)

Reply:
> Just seeing you asked the question yesterday. Then the easy thing that you can try is just ask permission to use in SciPy under our BSD-3-clause. Then in this code add a comment saying the code if from there and we got permission. (There are examples in our codebase you can look at where we did that if you want an example on how to write that.)

Is there any example for reference?

Reply:
```
Taken from https://stackoverflow.com/questions/2068372 by P.T. Roy,
written consent given on 23.04.2021 by the original author, Bruno Astrolino,
for free use in SciPy under the 3-clause BSD.
```

Reply:
I take it back even 

```python
len([x for x in shuffle(np.arange(13).tolist(), np.arange(13).tolist())])  # because it won't work with arrays
```
gives 10 million permutations. Are you sure this is a computable quantity in this way?

Reply:
Just wanted to flag this but even better then 👍 Thanks Ilhan!

Reply:
> I take it back even
> 
> ```python
> len([x for x in shuffle(np.arange(13).tolist(), np.arange(13).tolist())])  # because it won't work with arrays
> ```
> 
> gives 10 million permutations. Are you sure this is a computable quantity in this way?

Thanks. @ilayn @tupui That is the reason why in the original paper, the author does not choose a large array for demonstration. So in this case, I make two modes of estimation: "exact" and "asymptotic", in "exact" mode, we cannot scale to a large array since the permutation number goes crazy. But we can approximate the number by sampling, i.e "asymptotic" mode. In this mode, we can approximate the p-value in a good way without losing too much accuracy when the array goes large.

Reply:
Honestly, I probably would not bother implementing the exact method.

Reply:
> 

Yeah, but it is necessary for verification and benchmarking purpose.

Reply:
> e.g. in `stats._qmc.py`
> 
> ```
> Taken from https://stackoverflow.com/questions/2068372 by P.T. Roy,
> written consent given on 23.04.2021 by the original author, Bruno Astrolino,
> for free use in SciPy under the 3-clause BSD.
> ```

The stack exchange answerer replies to me and we can continue using this @tupui @ilayn, but definitely welcome any optimization!

Reply:
You can't benchmark anhthing concrete anyways so i would skip exact instead of pretending to have it implemented

Reply:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.permutation_test.html

Reply:
@mdhaber Nice shot, I did not realize there is ad hoc solution for permutation test, now I have integrated your solution to optimize the current bws test, take a look then!